### ISSUE TITLE
Redefinition error after installation of mss/msui

### ISSUE THREAD

**Issue by dhgkunkel** [Open-MSS/MSS#1641]
Hello there and happy new year!

I just started msui after a new installation (version 7.0.7) and I received a RefedinitionError.
Error message below:

> (mssenv) 10:25:16 dkunkel@arctic:~$ msui
> Traceback (most recent call last):
>   File "/usr/local/miniconda3/envs/mssenv/bin/msui", line 7, in <module> 
>     from mslib.msui.msui import main
>   File "/usr/local/miniconda3/envs/mssenv/lib/python3.10/site-packages/mslib/msui/msui.py", line 67, in <module>
>     from mslib.utils.qt import ui_mainwindow as ui
>   File "/usr/local/miniconda3/envs/mssenv/lib/python3.10/site-packages/mslib/utils/qt.py", line 639, in <module>
>     globals()[mod] = importlib.import_module("mslib.msui.qt5." + mod)
>   File "/usr/local/miniconda3/envs/mssenv/lib/python3.10/importlib/__init__.py", line 126, in import_module
>     return _bootstrap._gcd_import(name[level:], package, level)
>   File "/usr/local/miniconda3/envs/mssenv/lib/python3.10/site-packages/mslib/msui/qt5/ui_sideview_window.py", line 66, in <module>
>     from mslib.msui.mpl_qtwidget import MplSideViewWidget
>   File "/usr/local/miniconda3/envs/mssenv/lib/python3.10/site-packages/mslib/msui/mpl_qtwidget.py", line 46, in <module>
>     from mslib.utils.thermolib import convert_pressure_to_vertical_axis_measure
>   File "/usr/local/miniconda3/envs/mssenv/lib/python3.10/site-packages/mslib/utils/thermolib.py", line 31, in <module>
>     from mslib.utils.units import units, check_units
>   File "/usr/local/miniconda3/envs/mssenv/lib/python3.10/site-packages/mslib/utils/units.py", line 48, in <module>
>     units.define(pint.unit.UnitDefinition(_name, '', (),
>   File "/usr/local/miniconda3/envs/mssenv/lib/python3.10/site-packages/pint/registry.py", line 470, in define
>     self._define(definition)
>   File "/usr/local/miniconda3/envs/mssenv/lib/python3.10/site-packages/pint/registry.py", line 2124, in _define
>     definition, d, di = super()._define(definition)
>   File "/usr/local/miniconda3/envs/mssenv/lib/python3.10/site-packages/pint/registry.py", line 1459, in _define
>     definition, d, di = super()._define(definition)
>   File "/usr/local/miniconda3/envs/mssenv/lib/python3.10/site-packages/pint/registry.py", line 549, in _define
>     self._define_adder(d_def, d, di)
>   File "/usr/local/miniconda3/envs/mssenv/lib/python3.10/site-packages/pint/registry.py", line 557, in _define_adder
>     self._define_single_adder(
>   File "/usr/local/miniconda3/envs/mssenv/lib/python3.10/site-packages/pint/registry.py", line 579, in _define_single_adder
>     raise RedefinitionError(key, type(value))
> pint.errors.RedefinitionError: Cannot redefine 'sigma' (<class 'pint.definitions.UnitDefinition'>)

I had a similar error with version 7.0.5 which led to a new installation. However, in version 7.0.5 I received an attribute error:

> (mssenv) 10:17:59 dkunkel@arctic:~/.config/msui$ msui
> Traceback (most recent call last):
>   File "/usr/local/miniconda3/envs/mssenv/bin/msui", line 7, in <module>
>     from mslib.msui.msui import main
>   File "/usr/local/miniconda3/envs/mssenv/lib/python3.10/site-packages/mslib/msui/msui.py", line 67, in <module>
>     from mslib.utils.qt import ui_mainwindow as ui
>   File "/usr/local/miniconda3/envs/mssenv/lib/python3.10/site-packages/mslib/utils/qt.py", line 639, in <module>
>     globals()[mod] = importlib.import_module("mslib.msui.qt5." + mod)
>   File "/usr/local/miniconda3/envs/mssenv/lib/python3.10/importlib/__init__.py", line 126, in import_module
>     return _bootstrap._gcd_import(name[level:], package, level)
>   File "/usr/local/miniconda3/envs/mssenv/lib/python3.10/site-packages/mslib/msui/qt5/ui_sideview_window.py", line 66, in <module>
>     from mslib.msui.mpl_qtwidget import MplSideViewWidget
>   File "/usr/local/miniconda3/envs/mssenv/lib/python3.10/site-packages/mslib/msui/mpl_qtwidget.py", line 46, in <module>
>     from mslib.utils.thermolib import convert_pressure_to_vertical_axis_measure
>   File "/usr/local/miniconda3/envs/mssenv/lib/python3.10/site-packages/mslib/utils/thermolib.py", line 31, in <module>
>     from mslib.utils.units import units, check_units
>   File "/usr/local/miniconda3/envs/mssenv/lib/python3.10/site-packages/mslib/utils/units.py", line 48, in <module>
>     units.define(pint.unit.UnitDefinition(_name, '', (),
> AttributeError: module 'pint' has no attribute 'unit'. Did you mean: 'Unit'?

I wonder, if someone else has similar issues. 

Best, Daniel

**Comment by ReimarBauer** [comment]
please show the pint version e.g. 

mamba list pint

The fixation of pint  was done for 7.0.6 because of the failure you see in 7.0.5
https://github.com/Open-MSS/MSS/issues/1582

Have also a look if this works when you apply manually:
https://github.com/Open-MSS/MSS/pull/1584/files


If this solves it I can do tom. a backport and a release.

**Comment by dhgkunkel** [comment]
pint version:
> (mssenv) 12:03:00 dkunkel@arctic:~$ mamba list pint
> pint                      0.19.2             pyhd8ed1ab_0    conda-forge

The manual fix did not work, I still get the same error. Interestingly, it does only crash as 
`units.define("sigma = 1")` while I also included the line  `units.define("level = 1")` before. 

I comment the line:
`# units.define("sigma = 1")`
I get a totally different error:

> (mssenv) 12:09:11 dkunkel@arctic:~$ msui
> Traceback (most recent call last):
>   File "/usr/local/miniconda3/envs/mssenv/lib/python3.10/site-packages/fastkml/geometry.py", line 35, in <module>
>     from shapely.geometry import asShape
> ImportError: cannot import name 'asShape' from 'shapely.geometry' (/usr/local/miniconda3/envs/mssenv/lib/python3.10/site-packages/shapely/geometry/__init__.py)
> 
> During handling of the above exception, another exception occurred:
> 
> Traceback (most recent call last):
>   File "/usr/local/miniconda3/envs/mssenv/bin/msui", line 7, in <module>
>     from mslib.msui.msui import main
>   File "/usr/local/miniconda3/envs/mssenv/lib/python3.10/site-packages/mslib/msui/msui.py", line 71, in <module>
>     from mslib.msui import tableview, topview, sideview, linearview
>   File "/usr/local/miniconda3/envs/mssenv/lib/python3.10/site-packages/mslib/msui/topview.py", line 42, in <module>
>     from mslib.msui import kmloverlay_dockwidget as kml
>   File "/usr/local/miniconda3/envs/mssenv/lib/python3.10/site-packages/mslib/msui/kmloverlay_dockwidget.py", line 30, in <module>
>     from fastkml import kml, geometry, styles
>   File "/usr/local/miniconda3/envs/mssenv/lib/python3.10/site-packages/fastkml/__init__.py", line 30, in <module>
>     from .kml import KML, Document, Folder, Placemark
>   File "/usr/local/miniconda3/envs/mssenv/lib/python3.10/site-packages/fastkml/kml.py", line 37, in <module>
>     from .geometry import Geometry
>   File "/usr/local/miniconda3/envs/mssenv/lib/python3.10/site-packages/fastkml/geometry.py", line 42, in <module>
>     from pygeoif.geometry import as_shape as asShape
> ImportError: cannot import name 'as_shape' from 'pygeoif.geometry' (/usr/local/miniconda3/envs/mssenv/lib/python3.10/site-packages/pygeoif/geometry.py)
> CRITICAL:root:MSS Version: 7.0.7.
> CRITICAL:root:Python Version: 3.10.8 | packaged by conda-forge | (main, Nov 22 2022, 08:23:14) [GCC 10.4.0]
> CRITICAL:root:Platform: Linux-5.15.0-56-generic-x86_64-with-glibc2.35 (('64bit', 'ELF'))
> CRITICAL:root:Fatal error: Traceback (most recent call last):
>   File "/usr/local/miniconda3/envs/mssenv/lib/python3.10/site-packages/fastkml/geometry.py", line 35, in <module>
>     from shapely.geometry import asShape
> ImportError: cannot import name 'asShape' from 'shapely.geometry' (/usr/local/miniconda3/envs/mssenv/lib/python3.10/site-packages/shapely/geometry/__init__.py)
> 
> During handling of the above exception, another exception occurred:
> 
> Traceback (most recent call last):
>   File "/usr/local/miniconda3/envs/mssenv/bin/msui", line 7, in <module>
>     from mslib.msui.msui import main
>   File "/usr/local/miniconda3/envs/mssenv/lib/python3.10/site-packages/mslib/msui/msui.py", line 71, in <module>
>     from mslib.msui import tableview, topview, sideview, linearview
>   File "/usr/local/miniconda3/envs/mssenv/lib/python3.10/site-packages/mslib/msui/topview.py", line 42, in <module>
>     from mslib.msui import kmloverlay_dockwidget as kml
>   File "/usr/local/miniconda3/envs/mssenv/lib/python3.10/site-packages/mslib/msui/kmloverlay_dockwidget.py", line 30, in <module>
>     from fastkml import kml, geometry, styles
>   File "/usr/local/miniconda3/envs/mssenv/lib/python3.10/site-packages/fastkml/__init__.py", line 30, in <module>
>     from .kml import KML, Document, Folder, Placemark
>   File "/usr/local/miniconda3/envs/mssenv/lib/python3.10/site-packages/fastkml/kml.py", line 37, in <module>
>     from .geometry import Geometry
>   File "/usr/local/miniconda3/envs/mssenv/lib/python3.10/site-packages/fastkml/geometry.py", line 42, in <module>
>     from pygeoif.geometry import as_shape as asShape
> ImportError: cannot import name 'as_shape' from 'pygeoif.geometry' (/usr/local/miniconda3/envs/mssenv/lib/python3.10/site-packages/pygeoif/geometry.py)
> 
> QWidget: Must construct a QApplication before a QWidget
> Abgebrochen (Speicherabzug geschrieben)

**Comment by ReimarBauer** [comment]
geometry has more than one option  :(

```
try:
    from shapely.geometry import Point, LineString, Polygon
    from shapely.geometry import MultiPoint, MultiLineString, MultiPolygon
    from shapely.geometry.polygon import LinearRing
    # from shapely.geometry import GeometryCollection
    # Sean Gillies:
    # I deliberately omitted a geometry collection constructor because
    # there was almost no support in GEOS for operations on them. You
    # couldn't buffer a collection, for example, or find its difference
    # to another geometry. I've seen some signs of this changing in GEOS,
    # but until it does I don't think there's any point to the class.
    # It wouldn't be much more than a list of geometries.
    from pygeoif.geometry import GeometryCollection
    from shapely.geometry import asShape

except ImportError:
    from pygeoif.geometry import Point, LineString, Polygon
    from pygeoif.geometry import MultiPoint, MultiLineString, MultiPolygon
    from pygeoif.geometry import LinearRing
    from pygeoif.geometry import GeometryCollection
    from pygeoif.geometry import as_shape as asShape
```
shapely was also updated, 

try with sigma disabled and installing shapely version
`
mamba install shapely==1.8.5`


with the recent 2.0 asShape is removed see https://shapely.readthedocs.io/en/latest/migration.html?highlight=asShape#other-deprecated-functionality


the redefinition of sigma is solved in a newer pint version. We will see if I'm right with the next PR later today.


Instead of downgrading shapely you can also check that you have pygeoif==0.7. 

In my case I had 1.0, after downgrading to 0.7 it works too. We fixated fastkml==0.11. This has no fixation for pygeoif and uses the newest one. And shapely is not as a dependency defined, see bug report there. 

https://github.com/conda-forge/fastkml-feedstock/blob/main/recipe/meta.yaml#L24

I can do force both in our package, have to discuss this problem upstream

**Comment by ReimarBauer** [comment]
https://github.com/conda-forge/fastkml-feedstock/issues/7

**Comment by ReimarBauer** [comment]
btw. we catched the problem here too

 https://github.com/Open-MSS/MSS/actions/runs/3728746732/jobs/6324018066

unfortunately end of year ....

**Comment by ReimarBauer** [comment]
raise RedefinitionError(key, type(value)) is described at

https://github.com/hgrecco/pint/issues/1304

```
>>> from pint import UnitRegistry; ureg = UnitRegistry()
>>> ureg.parse_expression('sigma')
<Quantity(1, 'stefan_boltzmann_constant')>
```

**Comment by ReimarBauer** [comment]
We fixated fastkml==0.11 in the fastkml 0.12 pygeoif is fixated < 1. 
Because shapely is updated and got not working we have the wrong pygeoif too and it crashes.

**Comment by ReimarBauer** [comment]
The sigma "problem" occures now because of metpy==1.4.0 new release. It is undefined in older versions.

**Comment by ReimarBauer** [comment]
The dependencies will be fixed with build 2 on https://anaconda.org/conda-forge/mss/files

build 1 was missing a change in libtiff for *nix systems.

In ~1h after the files are uploaded the new build will be available. 

mamba search will show , e.g. 
7.0.7  py38h578d9bd_2  conda-forge  

The _2 is the build number.

With new builds I can adjust underlaying dependencies.

**Comment by ReimarBauer** [comment]
I close this after we released 7.0.8.
